projects/bootloader/test/test_bootloader_crc32.c

```diff
@@ -0,0 +1,54 @@
+#include <stdlib.h>
+
+#include "bootloader_crc32.h"
+#include "bootloader_mcu.h"
+#include "config.h"
+#include "crc32.h"
+#include "flash.h"
+#include "interrupt.h"
+#include "log.h"
+#include "persist.h"
+#include "test_helpers.h"
+#include "unity.h"
+
+uint32_t s_crc32_codes_size = 54;
+
+// intialize memory with random numbers for testing
+void intialize_memory() {
+  size_t curr_size = 0;
+  // writing the memory using a buffer with 2048 bytes at a time
+  uint8_t buffer[2048];
+
+  // generating random values in range of uint8(0->255)
+  for (int i = 0; i < 2048; i++) buffer[i] = rand() % (255 + 1 - 0) + 0;
+
+  while (curr_size <= BOOTLOADER_APPLICATION_SIZE) {
+    // write flash
+    flash_write((uintptr_t)BOOTLOADER_APPLICATION_START + curr_size, (uint8_t *)&buffer,
+                sizeof(buffer));
+    curr_size += sizeof(buffer);
+  }
+}
+
+void setup_test(void) {
+  flash_init();
+  crc32_init();
+  config_init();
+  intialize_memory();
+}
+
+void teardown_test(void) {}
+
+// test to check if computed crc32 code matches with the config
+// application_crc32
+void test_bootloader_application_crc32() {
+  BootloaderConfig config = { 0 };
+  config_get(&config);
+
+  // compute crc32 code
+  uint32_t computed_crc32 = calculated_application_crc32((uintptr_t)BOOTLOADER_APPLICATION_START,
+                                                         BOOTLOADER_APPLICATION_SIZE);
+
+  // fails, not sure if it is neccessary becuase we don't have the actual device?
+  // TEST_ASSERT_EQUAL(config.application_crc32, computed_crc32);
```

Comment:
this fails because the config crc32 is 0. Not sure if it is necessary to assert equal the codes here?

Reply:
The strategy for testing a crc32-like function like this would be to seed flash with data for which you know the crc32, then assert that the crc32 matches. In this case, using random data might not be the best idea - instead use something predictable (like set `buffer[i]` to `i % 231` or something). To find what crc32 it should have, I recommend creating the array in Python and passing it through the [zlib.crc32 function](https://docs.python.org/3/library/zlib.html#zlib.crc32) to find the crc32.

You don't need to do anything with `config` in this test, the config won't be set automatically or anything.

Reply:
done

Reply:
done